source/common/router/rds_impl.cc

```diff
@@ -80,24 +74,20 @@ RdsRouteConfigProviderImpl::RdsRouteConfigProviderImpl(
   config_source_ = MessageUtil::getJsonStringFromMessage(rds.config_source(), true);
 }
 
-RdsRouteConfigProviderImpl::~RdsRouteConfigProviderImpl() {
+RdsRouteConfigSubscription::~RdsRouteConfigSubscription() {
   // If we get destroyed during initialization, make sure we signal that we "initialized".
   runInitializeCallbackIfAny();
 
   // The ownership of RdsRouteConfigProviderImpl is shared among all HttpConnectionManagers that
   // hold a shared_ptr to it. The RouteConfigProviderManager holds weak_ptrs to the
   // RdsRouteConfigProviders. Therefore, the map entry for the RdsRouteConfigProvider has to get
   // cleaned by the RdsRouteConfigProvider's destructor.
-  route_config_provider_manager_.route_config_providers_.erase(manager_identifier_);
-}
-
-Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
-  return tls_->getTyped<ThreadLocalConfig>().config_;
+  route_config_provider_manager_.route_config_subscriptions_.erase(manager_identifier_);
```

Comment:
how do you avoid the race here?
an updated lister's rdsrouteconfig object is being deleted and tries to tell the routeconfig manager that it should be removed..
At the same time, an RDS update arrives and the subscription manager is updating all the routeconfig objects.

Reply:
There is not race.  ProviderImpl and RdsSubscription are separated.   ProviderImpl is tightly associated with ListenerImpl, one per one.  But RdsSubscription is shared, across ListenerImpl,  the ones with different names, or old one with the new one with config change.   A freed ProviderImpl will not free RdsSubscription if it is shared by others.  
BTW, here provider_manager should be called subscripton_manager.

Reply:
> BTW, here provider_manager should be called subscripton_manager.

provider_manager still manage providers (both static and rds), it doesn't own each rds provider but manages them through subscriptions.